### ISSUE TITLE
Unify configuration file formats

### ISSUE THREAD

**Issue by chrispcampbell** [climateinteractive/SDEverywhere#327]
Originally there was only one "config" file format used by SDE, which was the JSON "spec" file (e.g. `mymodel_spec.json`).  This file allows you to define which input and output variables are used in the generated C code, and supports a number of other things, like mappings/paths for DAT files, etc.

When I added support for the `sde dev` and `sde bundle` commands, I introduced a new "config" file format (`sde.config.js`) that provides a superset of the functionality of the original JSON "spec" files.  It serves a similar purpose, but allows for defining plugins and other things that require running code at build time.

Before we get to 1.0, it would be nice to resolve the differences between the two and clarify when to use one format over the other.  My thinking is that a JSON format is fine for simple use cases (using `sde generate`) where there is no need for plugins.  And the JS format is good for any use case (either `sde generate` or `sde {dev|bundle}`).

It would be nice if there was a single naming convention (for example, `sde.config.js[on]`) , and a smooth migration path if you want to start with JSON and work up to using JS, similar to how many other tools work in the JS ecosystem ([ESLint's config file formats](https://eslint.org/docs/latest/use/configure/configuration-files#configuration-file-formats) are one example).

**Comment by ToddFincannonEI** [comment]
The `spec.json` file can have any name that is given with the `--spec` or `-s` command-line argument. I agree that they can coexist as alternatives. That will be nice because I would not have to change the tooling for EPS.

**Comment by chrispcampbell** [comment]
This builds on the changes described in #765.  I decided to address both in the same branch since they go together.

---

TL;DR: There is no significant impact on how projects interact with the `ModelSpec` type exported by the build package.

Now that the compile package exports a `ModelSpec` type that specifies the fields that can appear in a `spec.json` file, I have made the build package depend on the compile package so that the `ModelSpec` and `ResolvedModelSpec` types in the build package can be based on the one from the compile package.  The only significant difference is that the build package `ModelSpec` type has two fields called `inputs` and `outputs` that accept objects, whereas the compile package `ModelSpec` type simply has `inputVarNames` and `outputVarNames` fields (since that package only works with the variable names).

The benefits of this approach are:
1. We now have the field names documented in one central place (in the compile package), and the build package simply re-exports those fields in its own types (no more duplication).
2. Users now have access to the same set of model spec options in the build and plugin-config packages (previously we had an untyped `options` field in the build package `ModelSpec` and `ResolvedModelSpec` interfaces, which was confusing).

(The `ModelSpec.options` field is still supported for now, but it is deprecated and may be removed in a future release.  I don't think many projects are using it currently anyway.)
